File: 003.Longest Substring Without Repeating Characters/c++/main.cpp
/**3. Longest Substring Without Repeating Characters
 * Given a string, find the length of the longest substring without repeating characters.

Examples:

Given "abcabcbb", the answer is "abc", which the length is 3.

Given "bbbbb", the answer is "b", with the length of 1.

Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.*/
#include <iostream>
using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)//滑动窗口
    {
        if(s == "") return 0;//特殊情况处理
        int l = 0;//左窗口
        int r = 0;//右窗口
        int ret = 1;//最长子串初始值
        int freq[256] = {0};//记录窗口中字符出现的次数
        freq[s[0]] = 1;//初始化
        while(l < s.size())
        {
            if(r+1 < s.size() && freq[s[r+1]] == 0)//右窗口还能扩展而且待扩展的位置还没有被使用过
                freq[s[++r]]++;//扩展右窗口，更新频率统计
            else//右窗口到底或者待扩展的位置与窗口中有冲突
                freq[s[l++]]--;//把左窗口收缩到冲突位置的右面
            ret = max(ret, (r-l+1));//ret值计算是贪心的，无论是右窗口到底还是产生了冲突都不会让ret减少
        }
        return ret;
    }
};

int main()
{
    string s = "abcabcbb";
    int res = Solution().lengthOfLongestSubstring(s);
    cout<<"res:"<<res<<endl;
    return 0;
}

File: 003.Longest Substring Without Repeating Characters/java/main.java
    class Solution {
        public int lengthOfLongestSubstring(String s) {
            //统计字母的频率，以字母的ascii码为下标
            int freq[] = new int[256];
            //当收缩左边界时会进行频度的减少，所以左边界的初始位置为0，即等待下标为0位置的字母被移除
            int l = 0;
            //当扩展右边界时才会进行频度的增加，所以右边界的初始位置为-1，即不包含任何字母
            int r = -1;
            //记录最长不含重复字母的子串长度
            int res = 0;
            //左边界未到达字符串尾部都是可以继续扩展的
            while(l < s.length()){
                //右边界向右扩展之后还没有出界而且新扩展位置的字母没有出现过
                //满足就向右扩展，并且更新频度
                if(r+1 < s.length() && freq[s.charAt(r+1)] == 0)
                    freq[s.charAt(++r)]++;
                //向右扩展之后越界或者是扩展位置字母已经存在了
                //满足就收缩左边界，并且更新频度
                else
                    freq[s.charAt(l++)]--;
                res = Math.max(res, r-l+1);
            }
            return res;
        }
    }